src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -937,6 +938,53 @@ class ObjectIdGenerator : public Generator<bsoncxx::types::b_oid> {
     const UniqueGenerator<std::string> _hexGen;
 };
 
+/**
+ * `{^UUID: {hex: "3b241101-e2bb-4255-8caf-4136c566a962"} }`
+ *  Generator for non-legacy UUIDs (0x04). hex can itself be a generator so the string can be
+ *  generated in many fashions ^Join, ^FormatString, ... etc.
+ *
+ *  The input format accepted for the hex field is 8-4-4-4-12 format. The code is agnostic as to the
+ *  UUID version.
+ *
+ *  see https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)
+ */
+class UuidGenerator : public Generator<bsoncxx::types::b_binary> {
+public:
+    UuidGenerator(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _node{node},
+          _hexGen{stringGenerator(node["hex"], generatorArgs)},
+          _subTypeGen{intGenerator(node["subType"],
+                                   generatorArgs,
+                                   static_cast<int64_t>(bsoncxx::binary_sub_type::k_uuid))} {}
+
+    bsoncxx::types::b_binary evaluate() override {
+        hex = _hexGen->evaluate();
+        return bsoncxx::types::b_binary{generateSubType(),
+                                        static_cast<uint32_t>(hex.length()),
+                                        reinterpret_cast<const uint8_t*>(hex.data())};
+    }
+
+private:
+    bsoncxx::binary_sub_type generateSubType() {
+        int64_t sub_type = _subTypeGen->evaluate();
+
+        if (static_cast<int64_t>(bsoncxx::binary_sub_type::k_uuid) != sub_type) {
+            std::stringstream msg;
+            msg << "Malformed binary sub_type. UUID only supports bsoncxx::binary_sub_type::k_uuid("
+                << static_cast<int64_t>(bsoncxx::binary_sub_type::k_uuid) << "), got " << sub_type;
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+        }
+        return static_cast<bsoncxx::binary_sub_type>(sub_type);
+    }
+
+    DefaultRandom& _rng;
+    const Node& _node;
+    const UniqueGenerator<std::string> _hexGen;
+    const UniqueGenerator<int64_t> _subTypeGen;
+    std::string hex;
```

Comment:
This bit took me the longest to figure out, without it the hex.data array in the returned ```bsoncxx::types::b_binary``` goes out of scope before the value is copied into the generated doc.

I'm happy to take a better approach here if there is one.